corehq/motech/openmrs/finders.py

```diff
@@ -0,0 +1,166 @@
+        self.property_weights = property_weights
+        self.threshold = threshold
+        self.confidence_margin = confidence_margin
+        self.property_map = {}
+
+    def set_property_map(self, case_config):
+        """
+        Set self.property_map to map OpenMRS properties and attributes
+        to case properties.
+        """
+        # Example value of case_config::
+        #
+        #     {
+        #       "person_properties": {
+        #         "birthdate": {
+        #           "doc_type": "CaseProperty",
+        #           "case_property": "dob"
+        #         }
+        #       },
+        #       // ...
+        #     }
+        #
+
+        for person_prop, value_source in case_config['person_properties'].items():
+            jsonpath = 'person.' + person_prop
+            self.property_map.update(get_caseproperty_jsonpathvaluemap(jsonpath, value_source))
+
+        for attr_uuid, value_source in case_config['person_attributes'].items():
+            jsonpath = 'person.attributes.value where `parent`.attributeType.uuid = "' + attr_uuid + '"'
+            self.property_map.update(get_caseproperty_jsonpathvaluemap(jsonpath, value_source))
+
+        for name_prop, value_source in case_config['person_preferred_name'].items():
+            jsonpath = 'person.preferredName.' + name_prop
+            self.property_map.update(get_caseproperty_jsonpathvaluemap(jsonpath, value_source))
+
+        for addr_prop, value_source in case_config['person_preferred_address'].items():
+            jsonpath = 'person.preferredAddress.' + addr_prop
+            self.property_map.update(get_caseproperty_jsonpathvaluemap(jsonpath, value_source))
+
+    def get_score(self, patient, case):
+        """
+        Return the sum of weighted properties to give an OpenMRS
+        patient a score of how well they match a CommCare case.
+        """
+        def weights():
+            for prop, weight in self.property_weights.items():
+                case_value = case.get_case_property(prop)
+                jsonpath_expr = parse(self.property_map[prop].jsonpath)
+                patient_value = jsonpath_expr.find(patient)
+                value_map = self.property_map[prop].value_map
+                is_equal = value_map.get(patient_value, patient_value) == case_value
+                yield weight if is_equal else 0
+
+        return sum(weights())
+
+    def find_patients(self, requests, case, case_config):
+        """
+        Matches cases to patients. Returns a list of patients, each
+        with a confidence score >= THRESHOLD
+        """
+        PatientScore = namedtuple('PatientScore', ['patient', 'score'])
+        self.set_property_map(case_config)
+
+        candidates = {}  # key on OpenMRS UUID to filter duplicates
+        for prop in self.searchable_properties:
+            value = case.get_case_property(prop)
+            response = requests.get('/ws/rest/v1/patient', {'q': value, 'v': 'full'})
```

Comment:
`/ws/rest/v1/patient` looks like a constant? Can just extract it out.

Reply:
Hmmm. That's true, [but][1] [there][2] [are][3] [a][4] [lot][5] [of][6] [endpoints][7] [like][8] [it][9] ... [and][10] [we][11] [only][12] [refer][13] to each of them once ... except that one, now that I'm going through them! I'll reuse [search_patients()][14], so at least all the endpoints are in the same module.


  [1]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L89
  [2]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L98
  [3]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L129
  [4]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L141
  [5]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L145
  [6]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L160
  [7]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L167
  [8]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L184
  [9]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L200
  [10]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L213
  [11]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L223
  [12]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L359
  [13]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L363
  [14]: https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/repeater_helpers.py#L166